client/protocols/guidance.py

```diff
@@ -0,0 +1,178 @@
+from client.protocols.common import models
+from client.utils import auth
+from urllib.request import urlopen
+import json
+import os
+import sys
+
+TGSERVER = "http://tg-server.app.cs61a.org/"
+LOCAL_TG = "tests/tg.ok_tg"
+OK_GUIDANCE = "tests/.ok_guidance"
+
+lambda_string_key_to_func = {
+    'none': lambda info, strMisU: None, 
+    'ki': lambda info, strMisU: info['ki'],
+    'reteach': lambda info, strMisU: info['dictMisU2ReteachMsg'].get(strMisU)
+}
+  
+# If student forces guidance messages to show, we will assign treatment group number below
+guidance_flag_dict = [1]
+
+class GuidanceProtocol(models.Protocol):
+    def __init__(self,working_dir,debug = -1):
+        self.tg_id = -1
+        self.msg_id = -1
+        self.misU_count = {}
+        self.working_dir = working_dir
+        self.debug = debug
+        try:
+            with open(working_dir + OK_GUIDANCE,"r") as f:
+                self.guidance_json = json.load(f)
+            self.error = False
+        except:
+            self.error = True
+
+    def guidance_msg(self,unique_id, input_lines,access_token,hash_key,guidance_flag):
+        if self.error:
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        shorten_unique_id = unique_id[unique_id.index('\n')+1:]
+        # Try to get th info dictionary for this question
+        wa_2_dict_info = self.guidance_json['dictAssessId2WA2DictInfo'].get(shorten_unique_id)
+
+        if not wa_2_dict_info: # get returns None if wa_2_dict_info doesn't have shorten_unique_id in dictionary
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        dict_info = wa_2_dict_info.get(repr(input_lines))
+
+        # If this wrong answer is not in the JSON file, display default message
+        if not dict_info:            
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+        if(self.debug != -1):
+            self.tg_id = self.debug
```

Comment:
Considering lines 54 and 55 are also setting the `tg_id` value. It would be clearer to move it to the `get_tg` function where all the other code that gets this value lives. In addition, since `get_tg` does not return anything but actually directly assigns `tg_id` a more accurate name would be `set_tg`. Another option would be to have `get_tg` actually return a value and lines 54-57 here are replaced with `self.tg_id = self.get_tg(access_token, guidance_flag)`.

Nitpick: Add a new line between 53 and 54 to make the code cleaner.